File: scripts/generate-data.py
#!/usr/bin/env python3

import argparse
import json
import ruamel.yaml

yaml = ruamel.yaml.YAML(typ="rt")
yaml.width = 4096
yaml.indent(mapping=2, sequence=4, offset=2)

parser = argparse.ArgumentParser(
    description="Generate completion and hover files from data",
    formatter_class=argparse.ArgumentDefaultsHelpFormatter,
)

parser.add_argument(
    "-i",
    dest="input_yaml",
    help="input yaml files",
    action="append",
    nargs="+",
    required=True,
)
parser.add_argument("--hover", dest="hover_file", help="output hover json", required=True)
parser.add_argument("--completion", dest="completion_file", help="output completion json", required=True)
parser.add_argument("--signature", dest="signature_file", help="output signatures json", required=False, default=None)
parser.add_argument(
    "--tooltip-lang", dest="tooltip_lang_id", help="language id used in intellisense tooltips", required=True
)

args = parser.parse_args()
input_yaml_list = args.input_yaml[0]
hover_file = args.hover_file
completion_file = args.completion_file
signature_file = args.signature_file
tooltip_lang_id = args.tooltip_lang_id


def load_data(yaml_list):
    result = {}
    for ypath in yaml_list:
        print(ypath)
        with open(ypath, encoding="utf8") as yaf:
            ydata = yaml.load(yaf)
        # merge, keep order
        result = {**result, **ydata}
    return result


def get_detail(item, include_types=True):
    """
    Returns full function invocation string: "int get_sfall_arg_at(int a, ObjectPtr b)"
    Types can be optionally omitted.
    """
    if "args" in item:
        # this is for hover and completion
        if include_types:
            args_with_type = [f"{a['type']} {a['name']}" for a in item["args"]]
            args_string = ", ".join(args_with_type)
            detail = f"{item['type']} {item['name']}({args_string})"
        else:  # and this is for signature
            args_no_type = [a["name"] for a in item["args"]]
            args_string = ", ".join(args_no_type)
            detail = f"{item['name']}({args_string})"
    else:
        if "detail" in item:
            detail = item["detail"]
        else:
            detail = item["name"]
    return detail


def get_doc(item):
    """
    Generates markdown doc from item args and doc field.
    """
    doc = ""
    if "args" in item:
        for arg in item["args"]:
            # doc = doc + "- `{} {}` {}\n".format(a["type"], a["name"], a["doc"])
            # type is already shown in detail
            doc = doc + f"- `{arg['name']}` {arg['doc']}\n"
    if "args" in item and "doc" in item:
        doc += "\n"
    if "doc" in item:
        doc = doc + item["doc"]
    return doc


def generate_completion(data):
    result = []
    COMPLETION_TAG_deprecated = 1  # pylint: disable=invalid-name
    for items_list in data:
        kind = data[items_list]["type"]
        items = data[items_list]["items"]
        for item in items:
            label = item["name"]

            completion_item = {"label": label, "kind": kind, "source": "builtin"}

            # CompletionItem "detail" field is a string in VScode and can't have highlighing
            # Instead, we add detail on top of documentation, which is MarkupContent
            detail = get_detail(item)
            doc = get_doc(item)
            if (detail != label) or (doc != ""):
                md_value = f"```{tooltip_lang_id}\n{detail}\n```"
                if doc != "":
                    md_value = f"{md_value}\n{doc}"
                markdown_contents = {"kind": "markdown", "value": md_value}
                completion_item["documentation"] = markdown_contents

            if "deprecated" in item:
                deprecated = item["deprecated"]
            else:
                deprecated = False
            if deprecated:
                completion_item["tags"] = [COMPLETION_TAG_deprecated]

            result.append(completion_item)
    return result


def generate_hover(data, lang_id):
    result = {}
    for items_list in data:
        items = data[items_list]["items"]
        for item in items:
            label = item["name"]

            # no point in showing a hover is there's no data besides the name
            if ("detail" not in item) and ("doc" not in item) and ("args" not in item):
                continue

            detail = get_detail(item)
            value = f"```{lang_id}\n{detail}\n```"
            doc = get_doc(item)
            if doc != "":
                value = f"{value}\n{doc}"

            markdown_contents = {"kind": "markdown", "value": value}
            hover = {"contents": markdown_contents}
            result[label] = hover
    return result


def markdown(string):
    """Returns VScode MarkupContent item"""
    return {"kind": "markdown", "value": string}


def generate_signatures(data, lang_id):
    sig_data = {}
    for items_list in data:
        items = data[items_list]["items"]
        for item in items:
            # signatures need args
            if "args" not in item:
                continue
            name = item["name"]

            label = get_detail(item, include_types=False)
            parameters = []
            for arg in item["args"]:
                doc = f"```{lang_id}\n{arg['type']} {arg['name']}\n```\n{arg['doc']}"
                parameters.append({"label": arg["name"], "documentation": markdown(doc)})

            function_doc = markdown(f"---\n{item['doc']}")
            sig_data[name] = {"label": label, "documentation": function_doc, "parameters": parameters}
    return sig_data


input_data = load_data(input_yaml_list)
completion_data = generate_completion(input_data)
hover_data = generate_hover(input_data, tooltip_lang_id)

if signature_file:
    signature_data = generate_signatures(input_data, tooltip_lang_id)

with open(hover_file, "w", encoding="utf8") as jf:
    json.dump(hover_data, jf, indent=4)

with open(completion_file, "w", encoding="utf8") as jf:
    json.dump(completion_data, jf, indent=4)

if signature_file:
    with open(signature_file, "w", encoding="utf8") as jf:
        json.dump(signature_data, jf, indent=4)

File: scripts/utils/src/generate-data.ts
/**
 * Generates LSP completion, hover, and signature JSON files from YAML data.
 * Replaces the former generate_data.py script.
 *
 * Usage:
 *   pnpm exec tsx scripts/utils/src/generate-data.ts \
 *     -i file1.yml file2.yml \
 *     --completion out/completion.json \
 *     --hover out/hover.json \
 *     --signature out/signature.json \
 *     --tooltip-lang fallout-ssl-tooltip
 */

import fs from "node:fs";
import { parseArgs } from "node:util";
import YAML from "yaml";
import { buildSignatureBlock, buildWeiduHoverContent, formatDeprecation } from "./shared/tooltip-format.js";
import { buildFalloutArgsTable, buildWeiduTable, type VarRow, type VarSection } from "./shared/tooltip-table.js";
import { WEIDU_TP2_CALLABLE_PREFIX } from "./shared/stanza-names.js";
import { WEIDU_JSDOC_TYPES } from "./shared/weidu-types.js";
import { cmpStr } from "./yaml-helpers.js";

// -- Types --

interface DataArg {
    readonly name: string;
    readonly type: string;
    readonly doc?: string;
    readonly required?: boolean;
    readonly default?: string;
}

interface DataItem {
    readonly name: string;
    readonly detail?: string;
    readonly doc?: string;
    readonly args?: readonly DataArg[];
    readonly rets?: readonly DataArg[];
    readonly type?: string;
    readonly deprecated?: boolean | string;
}

interface DataStanza {
    readonly type: number;
    readonly items: readonly DataItem[];
}

type DataFile = Record<string, DataStanza>;

interface MarkupContent {
    readonly kind: "markdown";
    readonly value: string;
}

interface CompletionResult {
    readonly label: string;
    readonly kind: number;
    readonly source: "builtin";
    readonly category: string;
    readonly documentation?: MarkupContent;
    readonly tags?: readonly number[];
    readonly detail?: string;
}

interface HoverResult {
    readonly contents: MarkupContent;
}

interface SignatureParam {
    readonly label: string;
    readonly documentation: MarkupContent;
}

interface SignatureResult {
    readonly label: string;
    readonly documentation: MarkupContent;
    readonly parameters: readonly SignatureParam[];
}

// -- Core logic (exported for testing) --

/** Known WeiDU item.type values that map to "X function NAME" detail strings. */
const WEIDU_ITEM_TYPES = new Set(["patch", "action", "dimorphic"]);

/**
 * Maps callable stanza names to their hover signature prefix.
 * Used to inject "action function", "patch macro", etc. into hover tooltips
 * at build time. Only applies when item has no structured args/rets
 * (items with args already get the prefix via getDetail's WeiDU path).
 */
const STANZA_CALLABLE_PREFIX = WEIDU_TP2_CALLABLE_PREFIX;

/**
 * Returns true if the item uses WeiDU-style structured format.
 * Detection: item.type is a known WeiDU callable type (patch/action/dimorphic)
 * or item has rets. Arg types alone are ambiguous ("int" appears in both Fallout and WeiDU).
 */
function isWeiduFormat(item: DataItem): boolean {
    if (item.type !== undefined && WEIDU_ITEM_TYPES.has(item.type)) {
        return true;
    }
    return item.rets !== undefined;
}

const COMPLETION_TAG_DEPRECATED = 1;

/**
 * Validates that parsed YAML data conforms to the DataFile structure.
 * Checks that each stanza has a numeric type and an items array.
 */
function validateDataFile(data: unknown, source: string): DataFile {
    if (typeof data !== "object" || data === null) {
        throw new Error(`Expected object in ${source}, got ${data === null ? "null" : typeof data}`);
    }
    const record = data as Record<string, unknown>;
    const result: Record<string, DataStanza> = {};
    for (const [key, value] of Object.entries(record)) {
        if (typeof value !== "object" || value === null) {
            throw new Error(`Expected object for stanza '${key}' in ${source}`);
        }
        const stanza = value as Record<string, unknown>;
        if (typeof stanza["type"] !== "number") {
            throw new Error(`Expected numeric 'type' in stanza '${key}' of ${source}`);
        }
        if (!Array.isArray(stanza["items"])) {
            throw new Error(`Expected 'items' array in stanza '${key}' of ${source}`);
        }
        result[key] = stanza as unknown as DataStanza;
    }
    return result;
}

function sortDataItems(items: readonly DataItem[]): readonly DataItem[] {
    return [...items].sort((a, b) => {
        const byName = cmpStr(a.name, b.name);
        if (byName !== 0) {
            return byName;
        }
        return cmpStr(a.detail ?? "", b.detail ?? "");
    });
}

function sortDataFile(data: DataFile): DataFile {
    const result: Record<string, DataStanza> = {};
    for (const stanzaName of Object.keys(data).sort(cmpStr)) {
        const stanza = data[stanzaName]!;
        result[stanzaName] = {
            ...stanza,
            items: sortDataItems(stanza.items),
        };
    }
    return result;
}

/**
 * Loads and merges multiple YAML data files.
 * Later files override earlier ones (last-writer-wins per stanza key).
 */
export function loadData(yamlPaths: readonly string[]): DataFile {
    let result: DataFile = {};
    for (const ypath of yamlPaths) {
        const content = fs.readFileSync(ypath, "utf8");
        const ydata = validateDataFile(YAML.parse(content) as unknown, ypath);
        result = { ...result, ...ydata };
    }
    return sortDataFile(result);
}

/**
 * Returns full function invocation string.
 * Fallout format: "int get_sfall_arg_at(int a, ObjectPtr b)"
 * WeiDU format: "dimorphic function SUBSTRING" (no args in signature)
 * When includeTypes is false, omits types (used for signature labels).
 *
 * @param stanzaName Optional stanza key (e.g., "action_functions"). When provided,
 *   items without structured args/rets get the callable prefix from STANZA_CALLABLE_PREFIX.
 */
export function getDetail(item: DataItem, includeTypes = true, stanzaName?: string): string {
    if (item.args !== undefined || item.rets !== undefined) {
        // WeiDU format: show "type function NAME" without parenthesized args
        if (isWeiduFormat(item)) {
            if (includeTypes && item.type !== undefined && WEIDU_ITEM_TYPES.has(item.type)) {
                return `${item.type} function ${item.name}`;
            }
            return item.name;
        }

        // Fallout format: show "returnType name(type arg, ...)"
        if (item.args !== undefined) {
            if (includeTypes) {
                const argsStr = item.args.map((a) => `${a.type} ${a.name}`).join(", ");
                return `${item.type} ${item.name}(${argsStr})`;
            }
            const argsStr = item.args.map((a) => a.name).join(", ");
            return `${item.name}(${argsStr})`;
        }
    }

    const base = item.detail ?? item.name;

    // For callable stanzas, prepend context+dtype prefix if not already present
    if (includeTypes && stanzaName) {
        const prefix = STANZA_CALLABLE_PREFIX[stanzaName];
        if (prefix && !base.startsWith(prefix)) {
            return `${prefix}${base}`;
        }
    }

    return base;
}

/** Fallout-style doc: 2-column table of args with descriptions + prose description. */
function getFalloutDoc(item: DataItem): string {
    let doc = "";
    if (item.args !== undefined) {
        const table = buildFalloutArgsTable(
            item.args.map((a) => ({ name: a.name, description: a.doc }))
        );
        if (table) {
            doc += table + "\n";
        }
    }
    if (item.args !== undefined && item.doc !== undefined) {
        doc += "\n";
    }
    if (item.doc !== undefined) {
        doc += item.doc;
    }
    return doc;
}

/** Map DataArg[] to VarRow[] for a given INT/STR category. */
function mapArgsToRows(args: readonly DataArg[], category: "int" | "str"): readonly VarRow[] {
    return args
        .filter((a) => WEIDU_JSDOC_TYPES.get(a.type)?.category === category)
        .map((a) => ({
            type: a.type,
            name: a.name,
            ...(a.required ? { required: true } : a.default !== undefined ? { default: a.default } : {}),
            ...(a.doc ? { description: a.doc } : {}),
        }));
}

/** Build the WeiDU param table string from a data item's args and rets. */
function getWeiduParamTable(item: DataItem): string {
    const sections: VarSection[] = [];

    if (item.args !== undefined && item.args.length > 0) {
        sections.push({ label: "INT vars", rows: mapArgsToRows(item.args, "int") });
        sections.push({ label: "STR vars", rows: mapArgsToRows(item.args, "str") });
    }

    if (item.rets !== undefined && item.rets.length > 0) {
        sections.push({
            label: "RET vars",
            rows: item.rets.map((r) => ({
                type: r.type,
                name: r.name,
                ...(r.doc ? { description: r.doc } : {}),
            })),
        });
    }

    return buildWeiduTable(sections);
}

function markdown(value: string): MarkupContent {
    return { kind: "markdown", value };
}

/**
 * Generates completion items from data stanzas.
 */
export function generateCompletion(data: DataFile, tooltipLangId: string): readonly CompletionResult[] {
    const sortedData = sortDataFile(data);
    // Count label occurrences across all stanzas
    const labelCounts = new Map<string, number>();
    for (const stanza of Object.values(sortedData)) {
        for (const item of stanza.items) {
            labelCounts.set(item.name, (labelCounts.get(item.name) ?? 0) + 1);
        }
    }

    const result: CompletionResult[] = [];
    for (const [stanzaName, stanza] of Object.entries(sortedData)) {
        const kind = stanza.type;
        for (const item of stanza.items) {
            const label = item.name;
            const detail = getDetail(item, true, stanzaName);

            let documentation: MarkupContent | undefined;
            if (isWeiduFormat(item)) {
                const paramTable = getWeiduParamTable(item);
                documentation = markdown(buildWeiduHoverContent({
                    signature: detail,
                    langId: tooltipLangId,
                    description: item.doc,
                    paramTable: paramTable || undefined,
                    deprecated: item.deprecated,
                }));
            } else {
                const doc = getFalloutDoc(item);
                if (detail !== label || doc !== "") {
                    let mdValue = buildSignatureBlock(detail, tooltipLangId);
                    if (doc !== "") {
                        mdValue += `\n${doc}`;
                    }
                    mdValue += formatDeprecation(item.deprecated);
                    documentation = markdown(mdValue);
                }
            }

            const deprecated = item.deprecated ?? false;
            const tags = deprecated ? [COMPLETION_TAG_DEPRECATED] as const : undefined;

            const completionItem: CompletionResult = {
                label,
                kind,
                source: "builtin",
                category: stanzaName,
                ...(documentation !== undefined ? { documentation } : {}),
                ...(tags !== undefined ? { tags } : {}),
                // Copy detail only for duplicate labels across stanzas.
                // Non-null assertion is safe: label comes from the same data that built labelCounts.
                ...(labelCounts.get(label)! > 1 ? { detail } : {}),
            };

            result.push(completionItem);
        }
    }
    return result;
}

/**
 * Generates hover data keyed by symbol name.
 */
export function generateHover(data: DataFile, langId: string): Record<string, HoverResult> {
    const sortedData = sortDataFile(data);
    const result: Record<string, HoverResult> = {};
    for (const [stanzaName, stanza] of Object.entries(sortedData)) {
        for (const item of stanza.items) {
            // Skip items with no data besides the name
            if (item.detail === undefined && item.doc === undefined && item.args === undefined && item.rets === undefined) {
                continue;
            }

            const label = item.name;
            const detail = getDetail(item, true, stanzaName);

            let value: string;
            if (isWeiduFormat(item)) {
                const paramTable = getWeiduParamTable(item);
                value = buildWeiduHoverContent({
                    signature: detail,
                    langId,
                    description: item.doc,
                    paramTable: paramTable || undefined,
                    deprecated: item.deprecated,
                });
            } else {
                value = buildSignatureBlock(detail, langId);
                const doc = getFalloutDoc(item);
                if (doc !== "") {
                    value += `\n${doc}`;
                }
                value += formatDeprecation(item.deprecated);
            }

            if (result[label]) {
                // Skip if identical content already present
                if (result[label]!.contents.value === value) {
                    continue;
                }
                // Merge overloads: append with separator
                result[label] = {
                    contents: markdown(result[label]!.contents.value + "\n\n---\n\n" + value),
                };
            } else {
                result[label] = { contents: markdown(value) };
            }
        }
    }
    return result;
}

/**
 * Generates signature help data for items with args.
 * WeiDU items include INT_VAR/STR_VAR category prefix in parameter docs.
 */
export function generateSignatures(data: DataFile, langId: string): Record<string, SignatureResult> {
    const sortedData = sortDataFile(data);
    const result: Record<string, SignatureResult> = {};
    for (const [, stanza] of Object.entries(sortedData)) {
        for (const item of stanza.items) {
            if (item.args === undefined) {
                continue;
            }

            const name = item.name;
            const weidu = isWeiduFormat(item);
            const label = getDetail(item, false);
            const parameters: SignatureParam[] = item.args.map((arg) => {
                const categoryPrefix = weidu
                    ? getCategoryPrefix(arg.type)
                    : "";
                const docStr = arg.doc ?? "";
                return {
                    label: arg.name,
                    documentation: markdown(`${buildSignatureBlock(`${categoryPrefix}${arg.type} ${arg.name}`, langId)}\n${docStr}`),
                };
            });

            const functionDoc = markdown(`---\n${item.doc ?? ""}`);
            result[name] = { label, documentation: functionDoc, parameters };
        }
    }
    return result;
}

/** Returns "INT_VAR " or "STR_VAR " prefix based on WeiDU type category. */
function getCategoryPrefix(type: string): string {
    const category = WEIDU_JSDOC_TYPES.get(type)?.category;
    if (category === "int") return "INT_VAR ";
    if (category === "str") return "STR_VAR ";
    return "";
}

// -- CLI entry point (tested via subprocess in generate-data-cli.test.ts) --

/* v8 ignore start -- CLI wrapper tested via execSync integration tests */
function main(): void {
    const { values } = parseArgs({
        options: {
            i: { type: "string", multiple: true, short: "i" },
            hover: { type: "string" },
            completion: { type: "string" },
            signature: { type: "string" },
            "tooltip-lang": { type: "string" },
        },
        strict: true,
    });

    const inputYaml = values.i;
    const hoverFile = values.hover;
    const completionFile = values.completion;
    const signatureFile = values.signature;
    const tooltipLangId = values["tooltip-lang"];

    if (inputYaml === undefined || hoverFile === undefined || completionFile === undefined || tooltipLangId === undefined) {
        console.error("Usage: generate-data -i <yaml...> --completion <path> --hover <path> --tooltip-lang <id> [--signature <path>]");
        process.exit(1);
    }

    const inputData = loadData(inputYaml);
    const completionData = generateCompletion(inputData, tooltipLangId);
    const hoverData = generateHover(inputData, tooltipLangId);

    fs.writeFileSync(hoverFile, JSON.stringify(hoverData, null, 4), "utf8");
    fs.writeFileSync(completionFile, JSON.stringify(completionData, null, 4), "utf8");

    if (signatureFile !== undefined) {
        const signatureData = generateSignatures(inputData, tooltipLangId);
        fs.writeFileSync(signatureFile, JSON.stringify(signatureData, null, 4), "utf8");
    }
}

const isDirectRun = process.argv[1]?.endsWith("generate-data.ts");
if (isDirectRun) {
    main();
}
/* v8 ignore stop */
